Add etcd-manager certificate expiration advisory (#9030)
kubernetes/kops: docs/advisories/etcd-manager-certificate-expiration.md

# etcd-manager Certificate Expiration

etcd-manager configures certificates for TLS communication between kube-apiserver and etcd, as well as between etcd members.
These certificates are signed by the cluster CA and are valid for a duration of 1 year.
Affed versions of etcd-manager did not handle the rotation of these certificates automatically.
This means if the certificates expire, the etcd cluster and kube-apiserver will be inaccessible.
This results in downtime of the cluster's control-plane.

## How do I know if I'm affected?

Clusters are affected by this issue if they're using a version of etcd-manager < XXXXXX.
The etcd-manager version is set automatically based on the Kops version.
These Kops versions are affected:

* Kops 1.10.0-alpha.1 through 1.16.1
* Kops 1.17.0-alpha.1 through 1.17.0-beta.1

> * Kops 1.18.0-alpha.1

Reviewer: https://github.com/kubernetes/kops/releases/tag/v1.18.0-alpha.2
```suggestion
* Kops 1.18.0-alpha.1 through 1.18.0-alpha.2
```

The issue can be confirmed by checking the ClusterSpec:

```

> kops get cluster $CLUSTER_NAME -o yaml

Reviewer: ```suggestion
kubectl get pods -n kube-system -l k8s-app=etcd-manager-main
```

Author: ahh thats much simpler :)

```

If the field `spec.etcdClusters[*].provider` is set to `manager`, or if it is not present and the cluster is managed with Kops >= 1.12.0, the cluster is affected.

## Solution

etcd-manager versions >= 3.0.20200428 will request new certificates before they expire.

Clusters can be upgraded to this version of etcd-manager in two ways:

> * Upgrade to Kops 1.16.2, 1.17.0-beta.2, or 1.18.1-alpha.2.

Reviewer: ```suggestion
* Upgrade to Kops 1.16.2, 1.17.0-beta.2, or 1.18.1-alpha.3.
```

This results in downtime of the cluster's control-plane.

## How do I know if I'm affected?

Clusters are affected by this issue if they're using a version of etcd-manager < XXXXXX.
The etcd-manager version is set automatically based on the Kops version.
These Kops versions are affected:

* Kops 1.10.0-alpha.1 through 1.16.1
* Kops 1.17.0-alpha.1 through 1.17.0-beta.1
* Kops 1.18.0-alpha.1 through 1.18.0-alpha.2

The issue can be confirmed by checking for the existence of etcd-manager 

`kubectl get pods -n kube-system -l k8s-app=etcd-manager-main -o jsonpath='{..image}' | tr -s '[[:space:]]' '\n'`

If this outputs `kopsio/etcd-manager` images with tags older than `3.0.20200428`, the cluster is affected.
If this outputs an image other than `kopeio/etcd-manager`, the cluster may be affected.
If this does does not output anything, the cluster is not affected.

## Solution

etcd-manager versions >= 3.0.20200428 will request new certificates before they expire.

Clusters can be upgraded to this version of etcd-manager in two ways. Both solutions require a rolling-update of the master nodes.

> * Upgrade to Kops 1.16.2, 1.17.0-beta.2, or 1.18.1-alpha.3.

Reviewer: ```suggestion
1. Upgrade to Kops 1.16.2, 1.17.0-beta.2, or 1.18.1-alpha.3.
```

  This is the recommended approach.
  Follow the normal steps when upgrading Kops and confirm the etcd-manager image will be updated based on the output of `kops update cluster`.

> * Override the etcd-manager image in the ClusterSpec:

Reviewer: Some people may skip some of the words when reading and do both things...
```suggestion

  2. Another approach is to override the etcd-manager image in the ClusterSpec:
```

1. Upgrade to Kops 1.16.2, 1.17.0-beta.2, or 1.18.1-alpha.3.
   This is the recommended approach.
   Follow the normal steps when upgrading Kops and confirm the etcd-manager image will be updated based on the output of `kops update cluster`.
   ```
   kops update cluster --yes
   kops rolling-update cluster --instance-group-roles=Master
   ```
2. Another solution is to override the etcd-manager image in the ClusterSpec:
   ```
   kops edit cluster $CLUSTER_NAME
   # Set `spec.etcdClusters[*].manager.image` to `kopeio/etcd-manager:3.0.20200428`
   kops update cluster # confirm the image is being updated
   kops update cluster --yes
   kops rolling-update cluster --instance-group-roles=Master --force
   ```

## Workaround

A rolling-update of the master nodes can be avoided by manually deleting the certificates to force them to be recreated.
This will not prevent the issue from occurring again, only reset the 1 year expiration.
Perform these steps on each master instance at a time.
1. SSH into the instance and delete the pki directory from each volume mount.
   ```

>    sudo rm -rf /mnt/master-vol-*/pki   # Mount paths may vary between cloud providers and OS distributions

Reviewer: Should make it easier to read in https://kops.sigs.k8s.io.
```suggestion
   # Mount paths may vary between cloud providers and OS distributions
   sudo rm -rf /mnt/master-vol-*/pki   
```

# etcd-manager Certificate Expiration

etcd-manager configures certificates for TLS communication between kube-apiserver and etcd, as well as between etcd members.
These certificates are signed by the cluster CA and are valid for a duration of 1 year.

> Affected versions of etcd-manager did not handle the rotation of these certificates automatically.

Reviewer 2: ```suggestion
Affected versions of etcd-manager currently do NOT automatically rotate these certificates before expiration.
```

> This means if the certificates expire, the etcd clusters and kube-apiserver will be inaccessible.
> This results in downtime of the cluster's control-plane.

Reviewer 2: ```suggestion
If these certificates are not rotated prior to their expiration, Kubernetes apiserver will become inaccessible and your control-plane will experience downtime.
```

## How do I know if I'm affected?

Clusters are affected by this issue if they're using a version of etcd-manager < `3.0.20200428`.
The etcd-manager version is set automatically based on the Kops version.
These Kops versions are affected:

> * Kops 1.10.0-alpha.1 through 1.16.1

Reviewer 3: ```suggestion
* Kops 1.16.0-alpha.1 through 1.16.1
```

* Kops 1.18.0-alpha.1 through 1.18.0-alpha.2

The issue can be confirmed by checking for the existence of etcd-manager pods and observing their image tags:

> 
> `kubectl get pods -n kube-system -l k8s-app=etcd-manager-main -o jsonpath='{..image}' | tr -s '[[:space:]]' '\n'`
> 

Reviewer: This should produce a slightly cleaner output. Also making code blocks as "bash" may help with formatting.
````suggestion

```bash
kubectl get pods -n kube-system -l k8s-app=etcd-manager-main \
  -o jsonpath='{range .items[*].spec.containers[*]}{.image}{"\n"}{end}'
```

````

Author: good idea 👍  I added that. Ideally we could get pods for both main and events clusters but there isn't a label that matches both sets of pods.

Reviewer: I don't think there is a case where they run different versions of `etcd-manager`. Best not tu confuse people.

Author: Each etcdCluster has its own Image field that could theoretically be overridden separately. It's probably not worth mentioning in this section but I am considering mentioning that in the 2nd Solution instructions.

* If this outputs `kopeio/etcd-manager` images with tags older than `3.0.20200428`, the cluster is affected.
* If this outputs an image other than `kopeio/etcd-manager`, the cluster may be affected.
* If this does does not output anything or outputs `kopeio/etcd-manager` images with tags >= `3.0.20200428`, the cluster is not affected.

## Solution

> etcd-manager versions >= `3.0.20200428` will request new certificates before they expire.

Reviewer 2: ```suggestion
Upgrade etcd-manager. etcd-manager version >= `3.0.20200428` manages certificate lifecycle and will automatically request new certificates before expiration
```

> Clusters can be upgraded to this version of etcd-manager in two ways. Both solutions require a rolling-update of the master nodes.

Reviewer 2: ```suggestion
We have two suggested workflows to upgrade etcd-manager in your cluster. While both workflows require a rolling-update of the master nodes, neither require control-plane downtime (if the clusters have highly available masters).
```

1. Upgrade to Kops 1.16.2, 1.17.0-beta.2, or 1.18.1-alpha.3.
   This is the recommended approach.
   Follow the normal steps when upgrading Kops and confirm the etcd-manager image will be updated based on the output of `kops update cluster`.
   ```
   kops update cluster --yes
   kops rolling-update cluster --instance-group-roles=Master
   ```
2. Another solution is to override the etcd-manager image in the ClusterSpec:
   ```
   kops edit cluster $CLUSTER_NAME
   # Set `spec.etcdClusters[*].manager.image` to `kopeio/etcd-manager:3.0.20200428`

>    kops update cluster # confirm the image is being updated

Reviewer 2: Do we _have_ to do a full cluster update? I can't piece together why you need to do a rolling-update in scenario #2.

Author: etcd-manager isnt a daemonset "bootstrap" manifest that gets applied during a `kops update cluster`, it is a static pod manifest created at instance-launch-time only. The only way for the pods to be replaced would be to replace the master instances, similar to the other static pods.

Reviewer 2: ahhh... that was why :) I forgot we were thinking about static pods. Carry on :+1:

   kops update cluster --yes
   kops rolling-update cluster --instance-group-roles=Master --force
   ```

> ## Workaround

Reviewer 2: I would call this something like "Hack/workaround" or "Danger" because this doesn't solve anyone's problems, just saves them for tomorrow.

A rolling-update of the master nodes can be avoided by manually deleting the certificates to force them to be recreated.

> This will not prevent the issue from occurring again, only reset the 1 year expiration.

Reviewer 2: Maybe move "This will not..." to the top of this section and bold it.

# etcd-manager Certificate Expiration

etcd-manager configures certificates for TLS communication between kube-apiserver and etcd, as well as between etcd members.
These certificates are signed by the cluster CA and are valid for a duration of 1 year.

Affected versions of etcd-manager currently do NOT automatically rotate these certificates before expiration.
If these certificates are not rotated prior to their expiration, Kubernetes apiserver will become inaccessible and your control-plane will experience downtime.

## How do I know if I'm affected?

Clusters are affected by this issue if they're using a version of etcd-manager < `3.0.20200428`.
The etcd-manager version is set automatically based on the Kops version.
These Kops versions are affected:

> * Kops 1.10.0-alpha.1 through 1.16.1

Reviewer 4: Safe versions are going to be 1.15.3, 1.16.2, 1.17.0-beta.2, 1.18.0-alpha.3

Releases of 1.15.3 and 1.16.2 are staged but not published (please take a look at the message).  I figured we should probably do them all at once...

Author: ok, i updated the list of affected versions and the list of fixed versions in the solutions section below.

Author: I updated the release drafts to link to kops.sigs.k8s.io, and confirmed this page looks good [on netlify](https://deploy-preview-9030--kubernetes-kops.netlify.app/advisories/etcd-manager-certificate-expiration/).

1. Upgrade to Kops 1.16.2, 1.17.0-beta.2, or 1.18.1-alpha.3.
   This is the recommended approach.
   Follow the normal steps when upgrading Kops and confirm the etcd-manager image will be updated based on the output of `kops update cluster`.
   ```
   kops update cluster --yes
   kops rolling-update cluster --instance-group-roles=Master
   ```
2. Another solution is to override the etcd-manager image in the ClusterSpec.
   The image will be set in two places, one for each etcdCluster (main and events).
   ```
   kops edit cluster $CLUSTER_NAME
   # Set `spec.etcdClusters[*].manager.image` to `kopeio/etcd-manager:3.0.20200428`
   kops update cluster # confirm the image is being updated
   kops update cluster --yes
   kops rolling-update cluster --instance-group-roles=Master --force
   ```

## Hack/Workaround

**This will not prevent the issue from occurring again, only reset the 1 year expiration.**

A rolling-update of the master nodes can be avoided by manually deleting the certificates to force them to be recreated.

> This will not prevent the issue from occurring again, only reset the 1 year expiration.

Reviewer 3: Since it was moved to the top of the section, perhaps remove this redundant copy?

Author: thanks, fixed
